internal/controllers/machinedeployment/machinedeployment_rolling.go

```diff
@@ -266,36 +310,44 @@ func (r *Reconciler) scaleDownOldMachineSetsForRollingUpdate(ctx context.Context
 
 	log.V(4).Info("Found available machines in deployment, scaling down old MSes", "count", availableMachineCount)
 
-	sort.Sort(mdutil.MachineSetsByCreationTimestamp(oldMSs))
+	sort.Sort(mdutil.MachineSetsByCreationTimestamp(p.oldMSs))
+
+	// TODO(in-place): fix this logic
+	//  It looks like that the current logic fails when the MD controller is called twice in a row e.g. reconcile of md 6 replicas MaxSurge=3, MaxUnavailable=1 and
+	//     ms1, 6/5 replicas << one is scaling down, but scale down not yet processed by the MS controller.
+	//     ms2, 3/3 replicas
+	//  Leads to:
+	//     ms1, 6/1 replicas << it further scaled down by 4, which leads to totAvailable machines is less than MinUnavailable, which should not happen
+	//     ms2, 3/3 replicas
 
 	totalScaledDown := int32(0)
 	totalScaleDownCount := availableMachineCount - minAvailable
-	for _, targetMS := range oldMSs {
-		if targetMS.Spec.Replicas == nil {
-			return 0, errors.Errorf("spec.replicas for MachineSet %v is nil, this is unexpected", client.ObjectKeyFromObject(targetMS))
-		}
-
+	for _, oldMS := range p.oldMSs {
 		if totalScaledDown >= totalScaleDownCount {
 			// No further scaling required.
 			break
 		}
 
-		if *(targetMS.Spec.Replicas) == 0 {
+		scaleIntent := ptr.Deref(oldMS.Spec.Replicas, 0)
```

Comment:
```suggestion
		oldMSReplicas := ptr.Deref(oldMS.Spec.Replicas, 0)
```
might be a better name